datasource/src/main/scala/quasar/physical/s3/impl/children.scala

```diff
@@ -77,57 +85,61 @@ object children {
         listingQuery
           .withQueryParam("prefix", prefix))
 
-      for {
-        // Send request to S3, parse the response as XML.
-        topLevelElem <- client.expect[xml.Elem](queryUri)
-        // Grab child object names from the response.
-        response <- for {
-          contents <- Sync[F].suspend {
-            // Grab <Contents>.
-            Sync[F].catchNonFatal((topLevelElem \\ "Contents"))
-              .adaptError {
-                case ex: Exception =>
-                  new Exception("XML received from AWS API has no top-level <Contents>", ex)
-              }
-          }
-
-          keyCount <- Sync[F].suspend {
-            // Grab <KeyCount> to ensure this response is not empty.
-            Sync[F].catchNonFatal((topLevelElem \\ "KeyCount").text.toInt)
-              .adaptError {
-                case ex: Exception =>
-                  new Exception("XML received from AWS API has no top-level <KeyCount>", ex)
-              }
-          }
-
-          // Grab all of the <Key> elements from <Contents>.
-          names <- contents.toList.traverse[F, String] { elem =>
-            Sync[F].catchNonFatal((elem \\ "Key").text)
-              .adaptError {
-                case ex: Exception =>
-                  new Exception("XML received from AWS API has no <Key> elements under <Contents>", ex)
-              }
-          }
-        } yield (names, keyCount)
-
-        (children, keyCount) = response
-
-        // Convert S3 object names to paths.
-        childPaths <- children.traverse(s3NameToPath)
-          .fold(Sync[F].raiseError[List[APath]](new Exception(s"Failed to parse object path in S3 API response")))(Sync[F].pure)
-
-        // Deal with some S3 idiosyncrasies.
-        // TODO: Pagination
-
-        result = if (keyCount === 0) {
-          None
-        } else {
-          // AWS includes the folder itself in the returned
-          // results, so we have to remove it.
-          // TODO: Report an error if there are duplicates. Remove duplicates.
-          Some(childPaths.filter(path => path =!= dir))
+    val request = Request[F](uri = queryUri)
+
+    for {
+      // Send request to S3, parse the response as XML.
+      signedRequest <- sign(request)
+
+      topLevelElem <- client.expect[xml.Elem](signedRequest)
+      // Grab child object names from the response.
+      response <- for {
+        contents <- Sync[F].suspend {
+          // Grab <Contents>.
+          Sync[F].catchNonFatal((topLevelElem \\ "Contents"))
```

Comment:
I think you actually want `\`. Using `\\` selects all `<Contents>` tags recursively at arbitrary depth.